source/link/linker.cpp

```diff
@@ -696,6 +729,57 @@ spv_result_t VerifyLimits(const MessageConsumer& consumer,
   return SPV_SUCCESS;
 }
 
+spv_result_t FixFunctionCallTypes(opt::IRContext& context,
+                                  const LinkageTable& linkings) {
+  auto mod = context.module();
+  const auto type_manager = context.get_type_mgr();
+  const auto def_use_mgr = context.get_def_use_mgr();
+
+  for (auto& func : *mod) {
+    func.ForEachInst([&](Instruction* inst) {
+      if (inst->opcode() != spv::Op::OpFunctionCall) return;
+      opt::Operand& target = inst->GetInOperand(0);
+
+      // only fix calls to imported functions
+      auto linking = std::find_if(
+          linkings.begin(), linkings.end(), [&](const auto& entry) {
+            return entry.exported_symbol.id == target.AsId();
+          });
+      if (linking == linkings.end()) return;
+
+      Function* exported_symbol_type =
+          type_manager->GetType(linking->exported_symbol.type_id)->AsFunction();
+
+      auto builder = InstructionBuilder(&context, inst);
+      for (uint32_t i = 1; i < inst->NumInOperands(); ++i) {
+        const Type* arg_type = exported_symbol_type->param_types()[i - 1];
+        if (arg_type->kind() != Type::kPointer) continue;
+
+        auto exported_func_param =
+            def_use_mgr->GetDef(linking->exported_symbol.parameter_ids[i - 1]);
+        const Type* target_type =
+            type_manager->GetType(exported_func_param->type_id());
+        if (target_type->kind() != Type::kPointer) continue;
```

Comment:
This `if` should never be taken, right? Cause the type of the parameter —when retrieved via the function's type— should be the same as when retrieved from the `OpFunctionParameter`.

There is nothing wrong with the code, but it would be worth adding a comment clarifying this is just a sanity check, in case of malformed input SPIR-V.

Reply:
Yeah.. `CheckImportExportCompatibility` should already fail if such a case would ever occur. I think I've added that before I've added the `Type::kPointer` checks in `CheckImportExportCompatibility`.

Anyway, I've removed it, because the SPIR-V would probably be invalid in either case.

Reply:
anyway, I've cleaned that part up a bit more, but also added a check to skip the bitcast if the type already matches with the current argument type.